File: solo/puzzles/easy/mime-type/mime_type.cpp
#include <iostream>
#include <map>
#include <string>

int main()
{
   std::map<std::string, std::string> mime;
   int n; // Number of elements which make up the association table.
   std::cin >> n; std::cin.ignore();
   int q; // Number Q of file names to be analyzed.
   std::cin >> q; std::cin.ignore();
   for (int i = 0; i < n; i++)
   {
      std::string ext; // file extension
      std::string mt; // MIME type.
      std::cin >> ext >> mt; std::cin.ignore();
      for (char& c : ext)
      {
         c = std::tolower(c);
      }
      
      if (mime.find(ext) == mime.end())
      {
         mime.insert({ ext, mt });
      }
   }
   for (int i = 0; i < q; i++)
   {
      std::string fname;
      std::getline(std::cin, fname); // One file name per line.
      size_t dot_position = fname.find_last_of('.');
      if (dot_position == std::string::npos)
      {
         std::cout << "UNKNOWN" << std::endl;
         continue;
      }

      std::string ext = fname.substr(dot_position + 1);
      for (char& c : ext)
      {
         c = std::tolower(c);
      }
      if (mime.find(ext) != mime.end())
      {
         std::cout << mime[ext] << std::endl;
      }
      else
      {
         std::cout << "UNKNOWN" << std::endl;
      }
   }
}

File: solo/puzzles/easy/mime-type/mime_type.rs
use std::io;
use std::collections::HashMap;

macro_rules! parse_input {
   ($x:expr, $t:ident) => ($x.trim().parse::<$t>().unwrap())
}

fn main() {
   let mut input_line = String::new();
   io::stdin().read_line(&mut input_line).unwrap();
   let n = parse_input!(input_line, i32); // Number of elements which make up the association table.
   let mut input_line = String::new();
   io::stdin().read_line(&mut input_line).unwrap();
   let q = parse_input!(input_line, i32); // Number Q of file names to be analyzed.
   let mut mime: HashMap<String, String> = HashMap::new();
   for i in 0..n as usize {
      let mut input_line = String::new();
      io::stdin().read_line(&mut input_line).unwrap();
      let inputs = input_line.split(" ").collect::<Vec<_>>();
      let ext = inputs[0].trim().to_string(); // file extension
      let mt = inputs[1].trim().to_string(); // MIME type.
      mime.insert(ext.to_lowercase(), mt);
   }
   for i in 0..q as usize {
      let mut input_line = String::new();
      io::stdin().read_line(&mut input_line).unwrap();
      let fname = input_line.trim_matches('\n').to_string(); // One file name per line.
      let ext = if fname.contains('.') {
         fname.split('.').last().unwrap_or("").to_lowercase()
      } else {
         String::new()
      };
      if let Some(mt) = mime.get(&ext) {
         println!("{}", mt);
      } else {
         println!("UNKNOWN");
      }
   }
}
